docs: added internal docs for solver flow (#4970)
garden-io/garden: core/src/graph/graph-execution-overview.md

* `StatusTaskNode`
    * When executed, calls the `getStatus` method of its underlying task.
    * Its dependencies are the status dependencies of the underlying task.
* `ProcessTaskNode`
    * When executed, calls the `process` method of its underlying task.
    * Its dependencies are the process dependencies of the underlying task.

Additionally, the solver maintains Maps of pending and in-progress nodes.

The flow is as follows:
* `garden.processTasks` is called with a list of tasks, which in turn calls `solver.solve`.
* The solver wraps each of these tasks in a `RequestTaskNode` and starts the loop (see the `loop` and `evaluateRequests` methods on the solver).
* In each iteration of the loop:
    * The solver looks at each incomplete request, checks if any of them were completed in the last iteration and completes them.
    * A status or process node may be added to the pending graph.
    * Note: Please refer to the source code for the specifics—any detailed description here is likely to drift from the implementation over time.
 and processes them in dependency order, making sure to first process any required dependencies.

> * The solver ensures that any remaining dependencies of `this.pendingNodes` are included.

Reviewer: ```suggestion
* The solver ensures that any remaining dependencies of its pending nodes (see `GraphSolver.pendingNodes`) are included.
```

* A dependency graph is generated from the remaining (i.e. not complete) pending nodes, and queried for the leaf nodes. This step also detects circular dependencies. 

> * Up to a certain concurrency limit, the leaf nodes of the dependency graph are processed (and added to the in-progress Map).

Reviewer: ```suggestion
* Up to a certain concurrency limit, the leaf nodes of the dependency graph are processed (and added to the in-progress Map, see `GraphSolver.inProgress`).
```

> * When a node finishes processing, it's marked ad complete, and the result of processing is saved on the node (this will be the return value of the `getStatus` or `process` method of the underlying task, for status and process nodes respectively`).

Reviewer: ```suggestion
* When a node finishes processing, it's marked ad complete, and the result of processing is saved on the node (this will be the return value of the `getStatus` or `process` method of the underlying task, for status and process nodes respectively).
```

* Eventually, all the requested tasks will be completed, and the results will be returned.

That's it! Of course, there are more details, but they mostly pertain to error handling and concurrency control, and are subject to change over time.

> What we've covered in this doc should serve as a decent introduction, but we highly reommend reading the source code. The best way to get a feel for the total flow is to break something, fix a bug or add a feature that touches on this part of Garden—diving in to the control flow with log statements or a debugger is the best way to get a feel for what's really going on.

Reviewer: ```suggestion
What we've covered in this doc should serve as a decent introduction, but we highly recommend reading the source code. The best way to get a feel for the total flow is to break something, fix a bug, or add a feature that touches on this part of Garden — diving into the control flow with log statements or a debugger is the best way to get a feel for what's really going on.
```

> Note: The `loop` and `ensurePendingNodes` methods on the solver are synchronous (i.e. not async) by design–by only updating the active nodes in synchronous code, we prevent race conditions and make the code much easier to reason about.

Reviewer: ```suggestion
Note: The `loop` and `ensurePendingNodes` methods on the solver are synchronous (i.e. not `async`) by design – by only updating the active nodes in synchronous code, we prevent race conditions and make the code much easier to reason about.
```
